### ISSUE TITLE
Additional mode for loading of segmentation data in tf.keras.preprocessing.image.ImageDataGenerator.flow_from_dataframe()

### ISSUE THREAD

**Issue by RobinBaumann** [tensorflow/tensorflow#37477]
**System information**
- TensorFlow version (you are using): 2.1.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

As of now, one has to come around with a workaround to load segmentation data in `tf.keras.preprocessing.image.ImageDataGenerator` to load images and segmentation masks in a generator. A current workaround could look like:
    
    datagen = tf.keras.preprocessing.image.ImageDataGenerator()
    
    df = pd.DataFrame({'images': images,
                       'masks': masks})
    # seed, so that image_generator and mask_generator will rotate and shuffle equivalently
    seed = 42
    image_generator = datagen.flow_from_dataframe(df, 
                                          directory='.', 
                                          x_col='images', 
                                          y_col='masks', 
                                          batch_size=1, 
                                          class_mode=None, 
                                          seed=seed)
    mask_generator = datagen.flow_from_dataframe(df, 
                                         directory='.', 
                                         x_col='masks', 
                                         y_col='images', # Or whatever 
                                         batch_size=1, 
                                         class_mode=None, 
                                         seed=seed)
    
    train_generator = zip(image_generator, mask_generator)
   
    # same for validation data generator

    model.fit(x=train_generator, epochs=EPOCHS, 
                    validation_data=validation_generator, use_multiprocessing=False)

It would be much cleaner, if there would be an option for `class_mode` (e.g. `class_mode='mask'`) that allows you to load the files specified by the paths in `y_col` as images. 

**Will this change the current api? How?**

An additional option for the argument `class_mode` will be added. 

**Who will benefit with this feature?**
Everyone who looks for a simple way to load images and corresponding segmentation data. 

**Any Other info.**

**Comment by RobinBaumann** [comment]
Okay, I tried to solve it on my own. However, there is still a dependency on [keras-preprocessing](https://github.com/keras-team/keras-preprocessing) that makes things a bit difficult. Are there any plans to integrate keras-preprocessing into TensorFlow?

**Comment by tanzhenyu** [comment]
Hi @RobinBaumann, do you have a more complete colab example or the stack trace so we know what is the dependency issue? We do have an equivalent `dataset_from_directory` but not ingesting dataframe for now.

**Comment by RobinBaumann** [comment]
In `tensorflow/python/keras/preprocessing/image.py` you can find an import statement to:

    from keras_preprocessing import image

Which is later used in:

    class ImageDataGenerator(image.ImageDataGenerator):
        ....

So all calls to the generator are handled by `keras_preprocessing.image.ImageDataGenerator` instead of `tf.keras.preprocessing.image.ImageDataGenerator`.  This is not an error, I just think that it would be cleaner to integrate all functionality `keras_preprocessing` into `tf.keras`, especially since Keras is already integrated in TF2 and will no longer be maintained standalone.

**Comment by tanzhenyu** [comment]
> In `tensorflow/python/keras/preprocessing/image.py` you can find an import statement to:
> 
> ```
> from keras_preprocessing import image
> ```
> 
> Which is later used in:
> 
> ```
> class ImageDataGenerator(image.ImageDataGenerator):
>     ....
> ```
> 
> So all calls to the generator are handled by `keras_preprocessing.image.ImageDataGenerator` instead of `tf.keras.preprocessing.image.ImageDataGenerator`, beause the latter doesn't provide implementations of `flow_from_dataframe`, `flow_from_directory` etc. This is not an error, I just think that it would be cleaner to integrate all functionality `keras_preprocessing` into `tf.keras`, especially since Keras is already integrated in TF2 and will no longer be maintained standalone.

OIC. Yeah what we currently have is dataset_from_directory, which is the equivalent of flow_from_directory, as experimental, we don't have a dataset_from_dataframe yet

**Comment by RobinBaumann** [comment]
> 
> OIC. Yeah what we currently have is dataset_from_directory, which is the equivalent of flow_from_directory, as experimental, we don't have a dataset_from_dataframe yet

Ah okay,  is currently anybody working on the `dataset_from_dataframe` or can this be considered a `good first issue` for me? :)

**Comment by tanzhenyu** [comment]
> > OIC. Yeah what we currently have is dataset_from_directory, which is the equivalent of flow_from_directory, as experimental, we don't have a dataset_from_dataframe yet
> 
> Ah okay, is currently anybody working on the `dataset_from_dataframe` or can this be considered a `good first issue` for me? :)

No one is working on that currently, you're welcome for contribution :-)

**Comment by tanzhenyu** [comment]
@RobinBaumann We're planning to get rid of keras-preprocessing dependency as well. If you want to contribute, that'd be great as well :-)

**Comment by tanzhenyu** [comment]
@RobinBaumann On another note, I would think it's probably not the best place to do the shuffle and rotate here. IMHO, it's better to separate the logic of 'load things from files' from 'preprocess those things', i.e.,:
step 1: load the image:
img = dataset_from_dataframe(...)
seg = dataset_from_dataframe(...)

step 2: preprocess it.
def map_fn(x):
  return tf.image.rotate(x, seed)  # Or something like that, we launched Keras Image Preprocessing layers.

preproc_img = img.shuffle(your_seed).map(map_fn)
preproc_seg = seg.shuffle(your_seed).map(map_fn)

**Comment by RobinBaumann** [comment]
> @RobinBaumann We're planning to get rid of keras-preprocessing dependency as well. If you want to contribute, that'd be great as well :-)

Sounds good 👍 I'm having a look into it this weekend and will come back to you, if i have some questions, if that's okay?

**Comment by tanzhenyu** [comment]
> > @RobinBaumann We're planning to get rid of keras-preprocessing dependency as well. If you want to contribute, that'd be great as well :-)
> 
> Sounds good 👍 I'm having a look into it this weekend and will come back to you, if i have some questions, if that's okay?

Of course!

**Comment by tanzhenyu** [comment]
I'm gonna close this for now, since we all agreed we should implement the dataset way. Please file a PR once it's ready. Thanks!
